Detectors/Vertexing/src/SVertexer.cxx

```diff
@@ -513,20 +546,38 @@ int SVertexer::checkCascades(float rv0, std::array<float, 3> pV0, float p2V0, in
       }
     }
     if (!goodHyp) {
+      LOG(debug) << "Casc not compatible with any hypothesis";
       continue;
     }
     auto& casc = mCascadesTmp[ithread].emplace_back(cascXYZ, pCasc, fitterCasc.calcPCACovMatrixFlat(candC), trNeut, trBach, mV0sTmp[ithread].size() - 1, bach.gid);
     o2::track::TrackParCov trc = casc;
     o2::dataformats::DCA dca;
-    if (!trc.propagateToDCA(pv, fitterCasc.getBz(), &dca, 5.) ||
+    if (!trc.propagateToDCA(cascPv, fitterCasc.getBz(), &dca, 5.) ||
         std::abs(dca.getY()) > mSVParams->maxDCAXYCasc || std::abs(dca.getZ()) > mSVParams->maxDCAZCasc) {
+      LOG(debug) << "Casc not compatible with PV";
+      LOG(debug) << "DCA: " << dca.getY() << " " << dca.getZ();
       mCascadesTmp[ithread].pop_back();
       continue;
     }
-    casc.setCosPA(cosPA);
-    casc.setVertexID(v0.getVertexID());
+
+    LOG(debug) << "Casc successfully added";
+    casc.setCosPA(bestCosPA);
+    casc.setVertexID(cascVtxID);
     casc.setDCA(fitterCasc.getChi2AtPCACandidate());
+
+    // clone the V0, set new cosPA and VerteXID, add it to the list of V0s
+    if (cascVtxID != v0.getVertexID()) {
+      auto v0clone = v0;
```

Comment:
This is not cloning: the clone should be added to the end of the  `mV0sTmp[ithread]`, i.e. 
```
auto& v0clone = mV0sTmp[ithread].push_back(v0);
```
But there is 1 more subtle point: some V0s are added only to check if there is a cascade for them, they are flagged by https://github.com/AliceO2Group/AliceO2/blob/5ce5ee631b75399add8a0869bf0edc8c8de99301/Detectors/Vertexing/src/SVertexer.cxx#L386. Imagine that you find a cascade for such a v0 but had to clone it because the cascade prefers different vertex. Then the v0 which should have been removed (since it has bad cosPA for the non-cascade V0 and is not used by any V0) will remain in the mV0sTmp[ithread].

I think at https://github.com/AliceO2Group/AliceO2/blob/5ce5ee631b75399add8a0869bf0edc8c8de99301/Detectors/Vertexing/src/SVertexer.cxx#L429 one should check if ant of the added cascades is using this V0 and if not, one should effectively copy the last added clone (in reality, redefine the vertexID and cosPA) to this v0, remove the clone and fix the reference from the cascade to removed cloned V0.

One more point :): you are cloning V0 for every cascade which prefers different vertex than original V0. In this case you may have multiple clones for the same V0/PV, while one should do this only once.

Reply:
> This is not cloning: the clone should be added to the end of the mV0sTmp[ithread], i.e.

But actually this is what I am doing: creating a clone here: 

> https://github.com/AliceO2Group/AliceO2/blob/5ce5ee631b75399add8a0869bf0edc8c8de99301/Detectors/Vertexing/src/SVertexer.cxx#L570

and pushing_back it here:

> https://github.com/AliceO2Group/AliceO2/blob/5ce5ee631b75399add8a0869bf0edc8c8de99301/Detectors/Vertexing/src/SVertexer.cxx#L576

or am I missing something?

-------------------------------------------------

> one should check if ant of the added cascades is using this V0 and if not, one should effectively copy the last added clone (in reality, redefine the vertexID and cosPA) to this v0

I see your point, working on it

-------------------------------------------------

> One more point :): you are cloning V0 for every cascade which prefers different vertex than original V0. In this case you may have multiple clones for the same V0/PV, while one should do this only once.

not sure I am getting your point. To me it looks that for every v0 you check for the cascade, and eventually you clone the v0 if the cascade is added with a different PV. But only the PV that guarantees the best CosPA is chosen. So in principle you can have at maximum two identical V0s w/ different PV. Isn't it?

Reply:
But what if you create a few cascades with the same V0 and PV different from the PV used by the V0? Would you add a V0 clone for each of these cascades, even though they use the same PV?

Sorry, you are right about cloning (missed `push_back` in  the end).